utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Code for the [`SinglePlaceholder`] pattern backend.
+
+use core::{cmp::Ordering, str::FromStr};
+use writeable::Writeable;
+
+use crate::common::*;
+use crate::Error;
+
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+
+/// A singleton enum for the [`SinglePlaceholder`] pattern backend.
+///
+/// # Examples
+///
+/// ```
+/// use core::cmp::Ordering;
+/// use icu_pattern::PatternItem;
+/// use icu_pattern::SinglePlaceholder;
+/// use icu_pattern::SinglePlaceholderKey;
+/// use icu_pattern::SinglePlaceholderPattern;
+///
+/// // Parse the string syntax and check the resulting data store:
+/// let pattern =
+///     SinglePlaceholderPattern::try_from_str("Hello, {0}!").unwrap();
+///
+/// assert_eq!(
+///     pattern.iter().cmp(
```

Comment:
did you add this for the use case where I want to get the individual parts out? It would be nice to have the length of this strongly typed.

Reply:
`pattern.iter()` is a core functionality, and it is a way to demonstrate where `SinglePlaceholder` is used in a public function which is why it is in this particular docs test.

The length isn't strongly typed. The iterator could have 0 (empty pattern), 1 (no literals or no element), 2 (element is at beginning or end), or 3 (element is in the middle) items.